helper/ssh/key_pair.go

```diff
@@ -0,0 +1,375 @@
+package ssh
+
+import (
+	"bytes"
+	"crypto/dsa"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"golang.org/x/crypto/ed25519"
+	gossh "golang.org/x/crypto/ssh"
+)
+
+const (
+	// That's a lot of bits.
+	defaultRsaBits = 4096
+
+	// Markers for various SSH key pair types.
+	Default KeyPairType = ""
+	Rsa     KeyPairType = "RSA"
+	Ecdsa   KeyPairType = "ECDSA"
+	Dsa     KeyPairType = "DSA"
+	Ed25519 KeyPairType = "ED25519"
+)
+
+// KeyPairType represents different types of SSH key pairs.
+// See the 'const' block for details.
+type KeyPairType string
+
+func (o KeyPairType) String() string {
+	return string(o)
+}
+
+const (
+	// UnixNewLine is a unix new line.
+	UnixNewLine NewLineOption = "\n"
+
+	// WindowsNewLine is a Windows new line.
+	WindowsNewLine NewLineOption = "\r\n"
+
+	// NoNewLine will not append a new line.
+	NoNewLine NewLineOption = ""
+)
+
+// NewLineOption specifies the type of new line to append to a string.
+// See the 'const' block for choices.
+type NewLineOption string
+
+func (o NewLineOption) String() string {
+	return string(o)
+}
+
+func (o NewLineOption) Bytes() []byte {
+	return []byte(o)
+}
+
+// KeyPairBuilder builds SSH key pairs.
+// It can generate new keys of type RSA and ECDSA.
+// It can parse user supplied keys of type DSA, RSA, ECDSA,
+// and ED25519.
+type KeyPairBuilder interface {
+	// SetType sets the key pair type.
+	SetType(KeyPairType) KeyPairBuilder
+
+	// SetBits sets the key pair's bits of entropy.
+	SetBits(int) KeyPairBuilder
+
+	// SetName sets the name of the key pair. This is primarily
+	// used to identify the public key in the authorized_keys file.
+	SetName(string) KeyPairBuilder
+
+	// SetPrivateKey takes an existing private key in PEM format.
+	// It overrides key generation details specified by SetType()
+	// and SetBits().
+	SetPrivateKey([]byte) KeyPairBuilder
+
+	// Build returns a SSH key pair.
+	//
+	// The following defaults are used if not specified:
+	//	Default type: ECDSA
+	//	Default bits of entropy:
+	//		- RSA: 4096
+	//		- ECDSA: 521
+	// 	Default name: (empty string)
+	Build() (KeyPair, error)
+}
+
+type defaultKeyPairBuilder struct {
```

Comment:
I think that by making the builder pattern you painted yourself into a complex corner. I also think the interfaces types aren't necessary.

1. `KeyPairBuilder` only has one implementer.
2. `KeyPair` has many implementers but they all are very similar. And their differences don't justify interfacing.
___

Now, I'll take a step back and see if I can come with something simpler.

Looking at [StepSshKeyPair](https://github.com/hashicorp/packer/blob/2ab2ea6ef131d3533f94a09759e8d5289a35c3e0/builder/virtualbox/common/step_ssh_key_pair.go#L31-L75rl). We need to:
1. parse the ssh key pair if there is one 
2. or create it
3. compute the public key from the private key
4. store it in the conf struct

Largely inspiring myself from [crypto/tls](https://github.com/golang/go/blob/0349f29a55fc194e3d51f748ec9ddceab87a5668/src/crypto/tls/generate_cert.go#L66-L92), here are a few funcs that should solve it:

```go
func generateKeyPair(ecdsaCurve string, rsaBits int) (*pem.Block, []byte, error) {
	var priv interface{}
	var err error
	switch ecdsaCurve {
	case "":
		priv, err = rsa.GenerateKey(rand.Reader, rsaBits)
	case "P224":
		priv, err = ecdsa.GenerateKey(elliptic.P224(), rand.Reader)
	case "P256":
		priv, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	case "P384":
		priv, err = ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
	case "P521":
		priv, err = ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
	default:
		return nil, nil, fmt.Errorf("Unrecognized elliptic curve: %q", ecdsaCurve)
	}
	if err != nil {
		return nil, nil, err
	}
	pem, err := pemBlockForKey(priv)
	if err != nil {
		return nil, nil, err
	}
	publicKey, err := publicKey(priv)
	return pem, publicKey, err
}

func pemBlockForKey(priv interface{}) (*pem.Block, error) {
	switch k := priv.(type) {
	case *rsa.PrivateKey:
		return &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(k)}, nil
	case *ecdsa.PrivateKey:
		b, err := x509.MarshalECPrivateKey(k)
		if err != nil {
			return nil, fmt.Errorf("Unable to marshal ECDSA private key: %v", err)
		}
		return &pem.Block{Type: "EC PRIVATE KEY", Bytes: b}, nil
	default:
		return nil, fmt.Errorf("Unrecognized key type: %T", priv)
	}
}

func publicKey(priv interface{}) ([]byte, error) {
	switch k := priv.(type) {
	case *rsa.PrivateKey, *ecdsa.PrivateKey:
		pk, err := gossh.NewPublicKey(k)
		if err != nil {
			return nil, err
		}
		return pk.Marshal(), nil
	case TODO: // this should probably have more cases !
	default:
		return nil, fmt.Errorf("Unrecognized key type: %T", priv)
	}
}
````

Reply:
This is probably something that @stephen-fox should weigh in on, but I have some thoughts...

- Isn't an advantage of interfacing that it invites future flexibility? Sure we only have one implementation now... Full disclosure: Just yesterday I ripped out an interface from a different project after having put it in there on advice from Steve ;)
- Putting this logic into StepSshKeyPair keeps it local to the virtualbox builder. I thought we wanted to make this capability available to all builders?
- That generateKeyPair function knows how to handle only two kinds of keys. We currently generate 2 types (omitted DSA because it's awful and ED25519 because of a clunky implementation issue) and we handle 4 types. What happens when the next key pair type (not RSA/ECDSA) is introduced? Isn't that flexibility an advantage? This function can't handle a not-yet-invented key type without breaking things. I think that's something Steve explicitly had in mind.

Reply:
Chris summarized pretty much everything I was going to say. However, I do want to add more back story explaining why we went in this direction. It is going to be kind of long... Sorry :(

When I started looking at the builder portion of the code base, I noticed that there was a lot of copy-pasted code. In particular, I noticed that the builder libraries do this with SSH key pairs code quite a bit. For example, `grep -r --exclude-dir vendor 'ecdsa.PrivateKey\|rsa.PrivateKey' .` indicates that about 20 libraries in the project are doing something with ECDSA and RSA key pairs.

I thought adding a library focused on creating SSH key pairs made the most sense at this point. I reasoned that eventually someone would have to do it anyways. I know creating a SSH key pair does not require a lot of code, but I would rather isolate that logic in a single library so I can easily test it and prove that the code actually works as desired, therefore giving others a reliable fixture to use in their code. This meant creating a basic, but flexible interface for creating and working with key pairs.

While this is definitely more complex, I think there is a use case for doing so. When Chris wanted to add support for building a key pair from an existing private key, he was able to do so by just adding to the `KeyPairBuilder` API - no public API breaking changes were required. Since unit tests were already in place, he was able to make the necessary changes and maintain confidence that if he broke the internal workings of the code, he would get called out for it.

In regards to each `KeyPair` implementation being very similar - I agree, but we had a good reason for doing this. Originally, there was only one implementation. However, Chris and I discovered a pretty bad bug during code review in `defaultKeyPairBuilder` (seen in d465231e633d). I forgot to set the ECDSA's curve, but I was still setting the bits of entropy in our `defaultKeyPair` struct to the correct value. As a result, the wrong curve was used, but the information about the curve's actual entropy was lost. This meant from a unit test's perspective, the key pair's entropy appeared correct, but the underlying key's entropy was wrong.

By having different implementations, we can store the original structs / interfaces that provide the actual answer to questions like "how many bits of entropy does this key pair have?". I think this might be beneficial down the line when users need more information about a key pair.

In hind sight - doing all of this added a lot of complexity to this change. Still, Chris and I think it needs to be done at some point, and it did not feel fair to copy paste something as important as generating a SSH key pair only to mangle to code to fit our needs.

Sorry for the small novel again.

Reply:
> Sorry for the small novel again.

No worries, thanks for taking the time to explain and bearing with me. 🙂  I'll try to answer with a not to long one !

___

> I thought adding a library focused on creating SSH key pairs made the most sense at this point.

Ah, sorry, I wasn't clear there; I totally agree on this !!! 🙂 This will centralise the calls and make this easier to maintain.

>  That generateKeyPair function knows how to handle only two kinds of keys.

Yes, there should be more type support, I mainly copy pasted this from `crypto/ssh`, I clearly didn't spend enough time making the example better, sorry. 🙂 

> When Chris wanted to add support for building a key pair from an existing private key, he was able to do so by just adding to the `KeyPairBuilder` API - no public API breaking changes were required.

> future flexibility

Yes, I can see that, but it's very unlikely that there will be a lot of new keypair types in the future. So if we want to support them all now. Shouldn't it be easy and done once ?
Also if we want to avoid a breaking change, the parameter can be a struct, ex:

```go
func GenerateKeyPair(c Config) (privateKey pem.Block, publicKey []byte, error)
```

( publicKey could also be of type `gossh.PublicKey` )

> Chris and I discovered a pretty bad bug during code review in `defaultKeyPairBuilder` (seen in [d465231](https://github.com/hashicorp/packer/commit/d465231e633d087d254f0bb47ff53190c8aef06c)) I forgot to set the ECDSA's curve, but I was still setting the bits of entropy in our `defaultKeyPair` struct to the correct value. As a result, the wrong curve was used [ ... ] This meant from a unit test's perspective, the key pair's entropy appeared correct, but the underlying key's entropy was wrong. By having different implementations, we can store the original structs / interfaces that provide the actual answer to questions like "how many bits of entropy does this key pair have?". I think this might be beneficial down the line when users need more information about a key pair.

Ah, I see. Well here, if there was an internal generate elliptic key func that takes a number and returns an `elliptic.Curve`, a unit test can solve this, ex : `if priv.PublicKey.Curve == elliptic.P224() {`. 

___

All in all, I still think this can be as nice **and** much more simple 🙂. I hope you agree.

Reply:
Another nice testing paradigm to have here would be to have the `rand.Reader` always return the same thing. To do so, random could be seeded with, say '1' in tests. That way, we could check in tests that generated key pairs are different/correct/etc.

Reply:
Thanks for the feedback. We will come up with something simpler.

Reply:
@azr, we refactored the SSH key pair code. It's down to two functions and three structs (two for creating, one for holding the resulting key pair data). Unit tests have been updated as well. Let us know what your thoughts are when you have a chance. We can make more changes as needed.

Reply:
Hey @stephen-fox, awesome ! Thanks for doing this 🙂. I'm looking into it right now.